### ISSUE TITLE
Moon libration

### ISSUE THREAD

**Issue by cosinekitty** [cosinekitty/astronomy#126]
Research feasibility of calculating libration of the Moon. Not sure how to even report that in a useful way. If I can find reliable test data and algorithms, and they do not add too much to the code size, I will implement this.

**Comment by cosinekitty** [comment]
Research is not very encouraging so far. Stellarium source code mentions libration only in regard to reading pre-computed values from the (very large) JPL ephemeris files. I still don't see any models for calculating libration.

**Comment by meshula** [comment]
Isn't it the case that you can compute the period of lunar libration relatively easily, but there's no models for magnitude of the libration?

**Comment by meshula** [comment]
(I implemented a simple model for the Apple Watch lunar face, but unfortunately I can't recall what I used, sorry for the vagueness.)

**Comment by cosinekitty** [comment]
@meshula Thanks for the pointer. I will take a look at that.

**Comment by meshula** [comment]
I don't recall what testing I did, but if I was going to do it today, I'd use the NASA or JPL tables, 2020 is available here https://svs.gsfc.nasa.gov/4768 and there seem to be breadcrumbs here https://naif.jpl.nasa.gov/pub/naif/generic_kernels/spk/planets/de430_moon_coord.pdf to the de430 ephemeris.

**Comment by cosinekitty** [comment]
Sweet! This one appears to be exactly what I need:
https://svs.gsfc.nasa.gov/vis/a000000/a004700/a004768/mooninfo_2020.txt
If I'm reading it right, the Elon and Elat columns are the libration angles.

**Comment by meshula** [comment]
Could probably verify that by playing with the interactive calculator?

**Comment by cosinekitty** [comment]
Yes, and I can also compare with the values in the Heavens Above moon page. It's great to have two independent data sources to corroborate the code's output.

**Comment by offsky** [comment]
I've been following along with this thread. Im glad that this looks possible.  From my comparison of the mooninfo_2020.txt, NASA dial a moon page, and heavens above it looks like there are 5 values that are important for libration. 
```
TXT		Dial-a-Moon			Heavens Above
------------------------------------------------------------------
Slon		Subsolar Longitude		?
Slat		Subsolar Latitude		?
Elon		Sub-Earth Longitude		Libration in longitude 
Elat		Sub-Earth Latitude		Libration in latitude 
AxisA		Position Angle			?
```

I am just learning this stuff, so I am probably wrong, but I think that if you have one of the lat/lon pairs you can calculate the other, and I think you can calculate the axis as well.  It would be nice if the function you write could return all 5 values so the developer can grab what they need.

Also, while Im here, it would be nice if it could also return the apparent size of the moon in arcseconds, which I realize isn't related to libration, so this might be another issue.

Im looking forward to this feature!

**Comment by cosinekitty** [comment]
I thought I'd drop by and let everyone know I have made some progress on libration.  I created a topic branch called `libration`. The commit 405a89fdf3dcf3aed11503bedc7aeb864a0596a7 in that branch implements PJ Naughter's algorithm in C. It calculates the libration angles in ecliptic longitude and ecliptic latitude. The results match NASA data within 0.13 arcminutes longitude and 1.67 arcminutes latitude. I'm trying to figure out why that second error is so high. So I'm still tinkering with it and there are a couple of things that might need fixing. I hope within the next week to finalize the C version, port it to the other 3 languages (C#, JavaScript, Python), then merge back into the `master` branch.
